flocker/node/test/test_deploy.py

```diff
@@ -1149,6 +1167,162 @@ def test_has_volume_no_changes(self):
+            applications=[application],
+        )
+        local_config = to_node(local_state).set(
+            applications=[application_with_larger_volume],
+            manifestations={manifestation.dataset_id: larger_manifestation},
+        )
+        assert_application_calculated_changes(
+            self, local_state, local_config, set(), no_change(),
+        )
+
+    def test_moved_volume_needs_changes(self):
+        """
+        If an ``Application`` is configured with a volume on a node but is no
+        longer configured to on that node, a change to stop that application is
+        calculated.
+        """
+        application = APPLICATION_WITH_VOLUME_SIZE
+        manifestation = application.volume.manifestation
+        local_state = EMPTY_NODESTATE.set(
+            devices={UUID(manifestation.dataset_id): FilePath(b"/dev/foo")},
+            paths={manifestation.dataset_id: FilePath(b"/foo/bar")},
+            manifestations={manifestation.dataset_id: manifestation},
+            applications=[application],
+        )
+        local_config = to_node(EMPTY_NODESTATE)
+        assert_application_calculated_changes(
+            self, local_state, local_config, set(),
+            sequentially(changes=[
+                in_parallel(changes=[
+                    StopApplication(application=application),
+                ]),
+            ]),
+        )
+
+    def test_different_volume_needs_change(self):
+        """
+        If an ``Application`` is configured with a volume but exists with a
+        different volume, a change to restart that application is calculated.
+        """
+        application = APPLICATION_WITH_VOLUME_SIZE
+        manifestation = application.volume.manifestation
+        another_manifestation = manifestation.transform(
+            ["dataset", "dataset_id"], uuid4(),
+        )
+        changed_application = application.transform(
+            ["volume", "manifestation"], another_manifestation,
+        )
+        local_state = EMPTY_NODESTATE.set(
+            devices={
+                UUID(manifestation.dataset_id): FilePath(b"/dev/foo"),
+                UUID(another_manifestation.dataset_id): FilePath(b"/dev/bar"),
+            },
+            paths={
+                manifestation.dataset_id: FilePath(b"/foo/bar"),
+                another_manifestation.dataset_id: FilePath(b"/bar/baz"),
+            },
+            manifestations={
+                manifestation.dataset_id: manifestation,
+                another_manifestation.dataset_id: another_manifestation,
+            },
+            applications=[application],
+        )
+        local_config = to_node(local_state).set(
+            applications=[
+                changed_application,
+            ],
+        )
+        assert_application_calculated_changes(
+            self, local_state, local_config, set(),
+            restart(application, changed_application, local_state),
+        )
+
```

Comment:
None of these tests actually cover the original reported problem: configuration specifies no size, actual dataset in state has one.